README.md
![Build status](https://secure.travis-ci.org/oreoshake/passw3rd.png)

It's only failing because it can't find the system ruby.  If you have a solution, checkout bin/passw3rd please :)

It will still work on your machine.

Introduction
------------------------------------------------------------------------------

This is a collection of encryption libraries intended to encrypt and store
passwords outside of source code.

Some advantages of keeping credentials out of source code are:

1. Credentials are not passed around when source code is shared.
2. Unintentional exposure of source code does not reveal credentials.
3. Read-access to source code can be much more permissive.
4. Source code can be checked into version control systems without concern
   for exposure of credentials.
5. It is easier to change credentials without having to worry about changing
   all instances.
6. Leaving credentials in source code leads to poor password management in
   general. If changing a credential requires you to change code, you are less
   likely to want to do it. 


Status
------------------------------------------------------------------------------

This project is IN PROGRESS. File bugs and feature requests.


License
------------------------------------------------------------------------------

License: MIT (see LICENSE file)


Credits
------------------------------------------------------------------------------

Copyright 2010, YELLOWPAGES.COM LLC
Development by Neil Matatall <neil.matatall@gmail.com>



History.txt
=== 0.2.1 / 2011-11-8

Added option to fail silently.  Pretty common need in a Rails environment.

=== 0.2.0 / 2011-11-5

Added acceptance tests as a contract for future features.  
Collapsed the KeyLoader class, eliminated the KeyPair class.  If for some reason you referenced those classes, things will break.  But you wouldn't do that, would you?
More testing goodness.

=== 0.1.3 / 2011-11-4

From key rotation, arose cipher changing

[passw3rd (gh-17-change_cipher *$)]$ rake change_cipher[~/passwords,~/passwords,aes-128-cbc,aes-256-cbc]
[passw3rd (gh-17-change_cipher *$)]$ rake change_cipher[~/passwords,~/passwords,aes-256-cbc,aes-256-cfb]

=== 0.1.2 / 2011-11-4

Added key rotation script
[passw3rd (gh-10-key_rotation *$)]$ rake rotate_keys[~/passwords,~/passwords,aes-256-cbc]

=== 0.1.0 / 2011-10-31

Finally some real life support
* Defaults to aes-256 for regulatory compliance (Thanks Edward Bonver)
* loads URIs, so password files can live anywhere (Thanks Ben Duong)
* Internals: code cleanup (Thanks @xternal), guard + test::unit + rspec love, better configuration options (thanks again @xternal)

=== 0.0.11 / 2011-10-12

Keys/ivs were being written as arrays, broke openssl command line compatibility:
$ openssl enc -e -aes-256-cbc -K `cat ~/.passw3rd-encryptionKey`  -iv `cat ~/.passw3rd-encryptionIV` -in README.md -out test.out
$ openssl enc -d -aes-256-cbc -K `cat ~/.passw3rd-encryptionKey`  -iv `cat ~/.passw3rd-encryptionIV` -out README.md -in test.out

=== 0.0.8 / 2011-10-02

Custom key directories

=== 0.0.6 / 2011-10-02

Added custom password directories, more tests.  Some code cleanup.  Java version maven site generation.

=== 0.0.1 / 2010-06-07

* 1 major enhancement

  * Adding ruby gem packaging



spec/config_spec.rb
describe "Config" do
  it "does stuff"
end

spec/spec_helper.rb
require 'passw3rd'

spec/password_service_spec.rb
require 'spec_helper'

describe ::Passw3rd::PasswordService do 
  it "restricts assigning ciphers to just a select few" do
    lambda { ::Passw3rd::PasswordService.cipher_name = "aes-256-donkey" }.should raise_error
  end
  
  describe "#_parse_uri" do
    it "detects uris" do
      ::Passw3rd::PasswordService._parse_uri("http://example.com").should be_is_a URI::HTTP
    end

    it "falls back to a file path" do
      URI.should_not_receive(:parse)
      ::Passw3rd::PasswordService._parse_uri("/some/file/some/where/over/the/rainbow").should be_is_a String
    end
  end
end

lib/passw3rd/key_loader.rb
require 'openssl'
module Passw3rd
  class KeyLoader
    KEY_FILE = "/.passw3rd-encryptionKey"
    IV_FILE = "/.passw3rd-encryptionIV"


    def self.load(path=nil) 
      if path.nil?
        path = ::Passw3rd::PasswordService.key_file_dir
      end

      begin
        key = IO.readlines(File.expand_path(path + KEY_FILE))[0]
        iv = IO.readlines(File.expand_path(path + IV_FILE))[0]
      rescue Errno::ENOENT
        puts "Couldn't read key/iv from #{path}.  Have they been generated?\n"
        raise $!
      end

      pair = KeyIVPair.new
      pair.key = [key].pack("H*")
      pair.iv = [iv].pack("H*")
      pair
    end

    def self.create_key_iv_file(path=nil)
      unless path
        path = ::Passw3rd::PasswordService.key_file_dir || ENV['HOME']
      end    

      # d'oh!
      cipher = OpenSSL::Cipher::Cipher.new(::Passw3rd::PasswordService.cipher_name)
      iv = cipher.random_iv
      key = cipher.random_key

      begin
        File.open(path + KEY_FILE, 'w') {|f| f.write(key.unpack("H*").join) }
        File.open(path + IV_FILE, 'w') {|f| f.write(iv.unpack("H*").join) }
      rescue
        puts "Couldn't write key/IV to #{path}\n"
        raise $!
      end
      path
    end
  end

  class KeyIVPair
    attr_accessor :key, :iv
  end
end

lib/passw3rd/password_client.rb
require 'optparse'

module Passw3rd
  class PasswordClient 
    password_file = nil
    gen_key_path = nil
    key_path = nil
    password_dir = nil
    mode = nil

    opts = OptionParser.new
    opts.banner = 'Usage: password_client [options]'

    opts.on('-d', '--decrypt PATH_TO_PASSWORD', 'Path to password file') do |opt|
      password_file = opt
      mode = "decrypt"
    end

    opts.on('-e', '--encrypt PASSWORD_FILE', 'Write the password to this location') do |opt|
      password_file = opt
      mode = "encrypt"
    end

    opts.on('-k', '--key-dir KEY_PATH', 'Use the keys specificed in this directory for encryption or decryption (default is ~/)') do |opt|
      key_path = opt
      unless File.directory?(File.expand_path(key_path))
        raise "#{opt} must be a directory"
      end
    end

    opts.on('-p', '--password-dir PATH', 'Read and write password files to this directory (default is ~/)') do |opt|
      password_dir = opt
      ::Passw3rd::PasswordService.password_file_dir = password_dir
      if !File.directory?(File.expand_path(password_dir))
        raise "#{password_dir} must be a directory"
      end      
    end

    opts.on('-g', '--generate-key [PATH]', 'generate key/iv and store in PATH, defaults to the home directory') do |opt|
      gen_key_path = opt
      if gen_key_path.nil?
        gen_key_path = ENV["HOME"]
      end
      if !File.directory?(File.expand_path(gen_key_path))
        raise "#{opt} is not a directory"
      end
    end

    opts.on_tail("-h", "--help", "Show this message") do
      puts opts
      exit
    end

    opts.parse(ARGV)

    # generate key/IV
    if gen_key_path
     ::Passw3rd::PasswordService.create_key_iv_file(gen_key_path)
      puts "generated keys in #{gen_key_path}"
    end

    # decrypt password_file using the key/IV in key_path
    if mode == "decrypt"
      decrypted =::Passw3rd::PasswordService.get_password(password_file, :key_path => key_path)
      puts "The password is: #{decrypted}"
    end

    # encrypt password, store it in output path
    if mode == "encrypt"
      begin
        system 'stty -echo'
        print "Enter the password: "
        password = STDIN.gets.chomp
      ensure
        system 'stty echo; echo ""'
      end

      file =::Passw3rd::PasswordService.write_password_file(password, password_file, key_path)
      puts "Wrote password to #{file}"
    end
  end
end

features/step_definitions/password_client_steps.rb
Given /^I have configured passw3rd to work in a sandbox$/ do
  ::Passw3rd::PasswordService.configure do |c|
    c.password_file_dir = Dir.tmpdir
    c.key_file_dir = Dir.tmpdir
    c.cipher_name = ::Passw3rd::APPROVED_CIPHERS.first
  end
end

Then /^my keys should be generated$/ do
  File.exists?(::Passw3rd::PasswordService.key_path).should be_true
  File.exists?(::Passw3rd::PasswordService.iv_path).should be_true
end

Given /^I have generated keys$/ do
  ::Passw3rd::PasswordService.create_key_iv_file
end

Then /^a password file for "([^"]*)" is created$/ do |file_name|
  # mv to helper
  path = File.join(::Passw3rd::PasswordService.password_file_dir, file_name)
  File.exists?(path).should be_true
end

Then /^the password file named "([^"]*)" should not contain "([^"]*)"$/ do |file_name, content|
  # mv to helper
  path = File.join(::Passw3rd::PasswordService.password_file_dir, file_name)  
  check_file_content(path, content, false)
end

Given /^I have a password file named "([^"]*)"(?: for the password "([^"]*)")?$/ do |file_name, password|
  password ||= "passw3rd"
  path = ::Passw3rd::PasswordService.write_password_file(password, file_name)
end

When /^I generate a password file named "([^"]*)"(?: for the password "([^"]*)")?$/ do |file_name, password|
  password ||= "passw3rd"
  steps %Q{
    When I run `passw3rd -e #{file_name}` interactively
    And I type "#{password}"
    And the output should contain "Wrote password to"
  }
end

When /^I generate keys$/ do
  steps %Q{
    When I successfully run `passw3rd -g`
  }
end

When /^I decrypt the password file named "([^"]*)"$/ do |file_name|
  steps %Q{
   When I successfully run `passw3rd -d #{file_name}`
  }
end
   




lib/passw3rd/password_service.rb
require 'open-uri'

module Passw3rd
  KEY_FILE = ".passw3rd-encryptionKey"
  IV_FILE = ".passw3rd-encryptionIV"
  # more preferred ciphers first
  APPROVED_CIPHERS = %w{aes-256-cbc aes-256-cfb aes-128-cbc aes-128-cfb}
  
  class PasswordService

    class << self
      attr_writer :password_file_dir
      def password_file_dir
        defined?(@password_file_dir) ? @password_file_dir : Dir.getwd
      end

      attr_writer :key_file_dir
      def key_file_dir
        defined?(@key_file_dir) ? @key_file_dir : Dir.getwd
      end

      def cipher_name= (cipher_name)
        raise "Hey man, you can only use #{APPROVED_CIPHERS}, you supplied #{cipher_name}" if cipher_name.nil? || !APPROVED_CIPHERS.include?(cipher_name)
        @cipher_name = cipher_name
      end

      def cipher_name
        defined?(@cipher_name) ? @cipher_name : APPROVED_CIPHERS.first
      end
    end

    def self.configure(&block)
      instance_eval &block
    end

    def self.get_password (password_file, options = {:key_path => self.key_file_dir, :force => true})
      uri = _parse_uri(password_file)
      encoded_password = Base64.decode64(open(uri) { |f| f.read })
      decrypt(encoded_password, options[:key_path])
    rescue => e
      raise ArgumentError, "Could not decrypt passw3rd file #{password_file} - #{e}" if options[:force]
    end

    def self.write_password_file(password, output_path, key_path = key_file_dir)
      enc_password = encrypt(password, key_path)
      base64pw = Base64.encode64(enc_password) 
      path = File.join(password_file_dir, output_path)
      open(path, 'w') { |f| f.write base64pw }
      path
    end

    def self.encrypt(password, key_path = key_file_dir)
      raise ArgumentError, "password cannot be blank" if password.to_s.empty?

      cipher = cipher_setup(:encrypt, key_path)
      begin
        e = cipher.update(password)
        e << cipher.final
      rescue OpenSSL::Cipher::CipherError => err
        puts "Couldn't encrypt password."
        raise err
      end
    end

    def self.decrypt(cipher_text, key_path = key_file_dir)
      cipher = cipher_setup(:decrypt, key_path)
      begin
        d = cipher.update(cipher_text)
        d << cipher.final
      rescue OpenSSL::Cipher::CipherError => err
        puts "Couldn't decrypt password #{cipher_text}.  Are you using the right keys (#{key_path})?"
        raise err
      end
    end

    def self.rotate_keys(args = {})
      unless args.empty?
        ::Passw3rd::PasswordService.configure do |c|
          c.password_file_dir = args[:password_file_dir]
          c.key_file_dir = args[:key_file_dir]
          c.cipher_name = args[:cipher]
        end
      end

      passwords = []

      Dir.foreach(::Passw3rd::PasswordService.password_file_dir) do |passw3rd_file|
        next if %w{. ..}.include?(passw3rd_file) || passw3rd_file =~ /\A\.passw3rd/
        puts "Rotating #{passw3rd_file}"
        passwords << {:clear_password => ::Passw3rd::PasswordService.get_password(passw3rd_file), :file => passw3rd_file}
      end

      ::Passw3rd::PasswordService.cipher_name = args[:new_cipher] if args[:new_cipher]

      path = self.create_key_iv_file
      puts "Wrote new keys to #{path}"

      passwords.each do |password|
        full_path = File.join(::Passw3rd::PasswordService.password_file_dir, password[:file])
        ::Passw3rd::PasswordService.write_password_file(password[:clear_password], password[:file])    
        puts "Wrote new password to #{full_path}"
      end
    end

    def self.create_key_iv_file(path=nil)
      unless path
        path = ::Passw3rd::PasswordService.key_file_dir || ENV['HOME']
      end    

      # d'oh!
      cipher = OpenSSL::Cipher::Cipher.new(::Passw3rd::PasswordService.cipher_name)
      iv = cipher.random_iv
      key = cipher.random_key

      begin
        File.open(key_path, 'w') {|f| f.write(key.unpack("H*").join) }
        File.open(iv_path, 'w') {|f| f.write(iv.unpack("H*").join) }
      rescue
        puts "Couldn't write key/IV to #{path}\n"
        raise $!
      end
      path
    end
    
    def self.key_path
      File.join(::Passw3rd::PasswordService.key_file_dir, KEY_FILE)
    end
    
    def self.iv_path
      File.join(::Passw3rd::PasswordService.key_file_dir, IV_FILE)
    end      

    protected
    
    def self.load_key(path=::Passw3rd::PasswordService.key_file_dir) 
      begin
        key = IO.readlines(File.expand_path(self.key_path))[0]
        iv = IO.readlines(self.iv_path)[0]
      rescue Errno::ENOENT
        puts "Couldn't read key/iv from #{path}.  Have they been generated?\n"
        raise $!
      end

      {:key => [key].pack("H*"), :iv => [iv].pack("H*")}
    end    

    def self.cipher_setup(method, key_path)
      pair = self.load_key(key_path)
      cipher = OpenSSL::Cipher::Cipher.new(cipher_name)
      cipher.send(method)
      cipher.key = pair[:key]
      cipher.iv = pair[:iv]
      cipher
    end

    def self._parse_uri password_file
      unless (password_file =~ URI::regexp(['ftp', 'http', 'https', 'file'])).nil?
        URI.parse(password_file)
      else
        File.join(password_file_dir, password_file)
      end
    end
  end
end
